rids/rids.py

```diff
@@ -15,113 +15,116 @@
 # limitations under the License.
 
 
+
 """
 Main entry point for RIDS, a Remote Intrusion Detection System.
 """
 
-
 import ipaddress
 import json
-import os
-import subprocess
-import sys
-import urllib.request
+import logging
+import threading
+import urllib
+import queue
 
 from absl import app
 from absl import flags
 
 from rids import network_capture
+from rids.iocs import iocs
 
 FLAGS = flags.FLAGS
 flags.DEFINE_string('host_ip', None,
                     'The IP of the host process, for ignoring direct requests')
-flags.DEFINE_string('config_path', 'config.json',
-                    'Path where IOC configuration can be found; uses config.json by default')
-
-
-def retrieve_bad_ips():
-  bad_ips = {}
-  with open(FLAGS.config_path, 'r') as f:
-    config = json.load(f)
-    for source in config['ioc_sources']:
-      # Download a fresh version of each IOC source in the config.
-      with urllib.request.urlopen(source['url']) as ioc_data:
-        # TODO check format of ioc source, branch to different parsers
-        # For now, the only source format is newline-seprated bad IPv4 addresses
-        # TODO perform parsing in a separate function from config handling
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', None,
+                    'file path indicating where to append new event logs')
 
 
 def main(argv):
+  """Main entry point of the app.  Also bound to CLI `rids` by setuptools."""
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  if FLAGS.eventlog_path:
+    logging.basicConfig(level=logging.DEBUG,
+                        format='%(asctime)s %(levelname)-8s %(message)s',
+                        datefmt='%m-%d %H:%M',
+                        filename=FLAGS.eventlog_path,
+                        filemode='a')
+
+  config = _load_config()
+  ruleset = iocs.fetch_iocs(config)
+
+  host_ip = _get_host_ip()
+  print(f'Using {host_ip} as host IP address')
+
+  event_queue = queue.Queue()
+  threading.Thread(
+      target=_inspect_tls_traffic,
+      args=(host_ip, ruleset, event_queue))
+  threading.Thread(
```

Comment:
If one of the threads fail, can we make the whole program fail?

Reply:
Let's file a bug/feature for that and do it in another PR.  I agree we can add that (or, rather, restart the failed thread maybe? it is usually easier to restart than recover, though, and systemctl would restart the process, so failing sgtm).  But it looks like I'd have to subclass Thread and re-define the run() and join() methods (and a significant change to main since we don't ever join() on the threads, so putting the queue reading in another thread).. or use sys.excepthook but that's only available from python 3.8+.  Maybe there's an alternative way that's better, I'll look into it.  We could add something that watches the health of the tshark processes but don't really need to since it would close the pipe and finish the thread, but we should react to that happening by failing the process too.

Reply:
Sounds good.

Reply:
asyncio may help: https://docs.python.org/3/library/asyncio-subprocess.html